lib/lightning_web/controllers/api/workflows_controller.ex

```diff
@@ -0,0 +1,461 @@
+    project = Repo.get(Project, project_id)
+
+    Permissions.can(
+      Lightning.Policies.Workflows,
+      access,
+      conn.assigns.current_resource,
+      project
+    )
+  end
+
+  defp maybe_handle_error(conn, result, workflow_id \\ nil)
+
+  defp maybe_handle_error(_conn, result, _workflow_id) when is_map(result),
+    do: result
+
+  defp maybe_handle_error(conn, {:error, :conflict, name}, workflow_id),
+    do:
+      conn
+      |> put_status(:conflict)
+      |> json(%{
+        id: workflow_id,
+        errors: %{
+          workflow: [
+            "Cannot save a workflow (#{name}) while it is being edited on the App UI"
+          ]
+        }
+      })
+
+  defp maybe_handle_error(conn, {:error, :invalid_id, id}, workflow_id),
+    do:
+      reply_422(
+        conn,
+        workflow_id,
+        :workflow,
+        "Id #{id} should be a UUID."
+      )
+
+  defp maybe_handle_error(conn, {:error, :invalid_workflow_id}, workflow_id),
+    do:
+      reply_422(
+        conn,
+        workflow_id,
+        :workflow,
+        "Edges, jobs and triggers cannot reference another workflow!"
+      )
+
+  defp maybe_handle_error(conn, {:error, :missing_id}, workflow_id),
+    do:
+      reply_422(
+        conn,
+        workflow_id,
+        :workflow,
+        "All jobs and triggers should have an id (UUID)."
+      )
+
+  defp maybe_handle_error(conn, {:error, {:duplicated_ids, ids}}, workflow_id),
+    do:
+      reply_422(
+        conn,
+        workflow_id,
+        :workflow,
+        "The ids #{inspect(ids)} should be unique for all workflows."
+      )
+
+  @reason_entity_field %{
+    invalid_triggers: {"Trigger", :triggers},
+    invalid_jobs: {"Job", :jobs},
+    invalid_edges: {"Edge", :edges}
+  }
+  defp maybe_handle_error(
+         conn,
+         {:error, {reason, id_to_errors_map}},
+         workflow_id
+       )
+       when reason in [:invalid_triggers, :invalid_jobs, :invalid_edges] do
+    {entity, workflow_field} = Map.get(@reason_entity_field, reason)
+
+    error_msgs =
+      id_to_errors_map
+      |> Enum.map(fn {id, errors} ->
+        errors =
+          Enum.map(errors, fn
+            {field, [error]} -> "#{field} #{error}"
+            {field, errors} -> "#{field} #{inspect(errors)}"
+          end)
+
+        "#{entity} #{id} has the errors: [#{errors}]"
+      end)
+
+    reply_422(
+      conn,
+      workflow_id,
+      workflow_field,
+      error_msgs
+    )
+  end
+
+  defp maybe_handle_error(conn, result, workflow_id)
+       when is_tuple(result) do
+    case result do
+      {:error, :invalid_jobs_ids, job_ids} ->
+        reply_422(
+          conn,
+          workflow_id,
+          :jobs,
+          "The jobs #{inspect(job_ids)} should be present both in the jobs and on an edge."
+        )
+
+      {:error, :edges_misses_a_trigger} ->
+        reply_422(
+          conn,
+          workflow_id,
+          :edges,
+          "Missing edge with source_trigger_id."
```

Comment:
This error occurs when an edge points to a source trigger that does not exist. But the error implies that it's the edge that is missing

Maybe better wording is `Trigger not found for source_trigger_id`